isobus/src/isobus_functionalities.cpp

```diff
@@ -0,0 +1,872 @@
+			{
+				retVal = 14;
+			}
+			break;
+
+			default:
+			{
+				// Option not handled. This is normal for NoOption but others should be considered.
+			}
+			break;
+		}
+		return retVal;
+	}
+
+	std::uint8_t ControlFunctionFunctionalities::get_tim_option_bit_index(TractorImplementManagementOptions option) const
+	{
+		std::uint8_t retVal = 0xFF;
+
+		switch (option)
+		{
+			case TractorImplementManagementOptions::FrontPTODisengagementIsSupported:
+			case TractorImplementManagementOptions::RearPTODisengagementIsSupported:
+			case TractorImplementManagementOptions::FrontHitchMotionIsSupported:
+			case TractorImplementManagementOptions::RearHitchMotionIsSupported:
+			case TractorImplementManagementOptions::VehicleSpeedInForwardDirectionIsSupported:
+			case TractorImplementManagementOptions::GuidanceCurvatureIsSupported:
+			{
+				retVal = 0;
+			}
+			break;
+
+			case TractorImplementManagementOptions::FrontPTOEngagementCCWIsSupported:
+			case TractorImplementManagementOptions::RearPTOEngagementCCWIsSupported:
+			case TractorImplementManagementOptions::VehicleSpeedInReverseDirectionIsSupported:
+			{
+				retVal = 1;
+			}
+			break;
+
+			case TractorImplementManagementOptions::FrontHitchPositionIsSupported:
+			case TractorImplementManagementOptions::RearHitchPositionIsSupported:
+			case TractorImplementManagementOptions::VehicleSpeedStartMotionIsSupported:
+			{
+				retVal = 2;
+			}
+			break;
+
+			case TractorImplementManagementOptions::FrontPTOengagementCWIsSupported:
+			case TractorImplementManagementOptions::RearPTOEngagementCWIsSupported:
+			case TractorImplementManagementOptions::VehicleSpeedStopMotionIsSupported:
+			{
+				retVal = 3;
+			}
+			break;
+
+			case TractorImplementManagementOptions::FrontPTOspeedCCWIsSupported:
+			case TractorImplementManagementOptions::RearPTOSpeedCCWIsSupported:
+			case TractorImplementManagementOptions::VehicleSpeedForwardSetByServerIsSupported:
+			{
+				retVal = 4;
+			}
+			break;
+
+			case TractorImplementManagementOptions::FrontPTOspeedCWIsSupported:
+			case TractorImplementManagementOptions::RearPTOSpeedCWIsSupported:
+			case TractorImplementManagementOptions::VehicleSpeedReverseSetByServerIsSupported:
+			{
+				retVal = 5;
+			}
+			break;
+
+			case TractorImplementManagementOptions::VehicleSpeedChangeDirectionIsSupported:
+			{
+				retVal = 6;
+			}
+			break;
+
+			default:
+			{
+				// Option not handled. This is normal for NoOption but others should be considered.
+			}
+			break;
+		}
+		return retVal;
+	}
+
+	bool ControlFunctionFunctionalities::pgn_request_handler(std::uint32_t parameterGroupNumber,
+	                                                         ControlFunction *,
+	                                                         bool &acknowledge,
+	                                                         AcknowledgementType &,
+	                                                         void *parentPointer)
```

Comment:
The `parentPointer` is of type `void *` here, wouldn't it make more sense if we just specify the type we want the pointer to point to. E.g. `ControlFunctionFunctionalities *` here?

I'm not sure why we decided to use these `void *` pointers when using them internally in private/protected functions for parents? But I feel like I prefer overloading the function if we need to support pointers to multiple types of parents. I know we use it in more places throughout the project but I just now begin to question if it's optimal choice for readability/maintainability. What are your thoughts on this?

Reply:
Oh, and we will make Sonar happier as well by specifying a meaningful type :)

Reply:
In this case it's because we're matching the signature of the callback inside `ParameterGroupNumberRequestProtocol` which doesn't know anything about `ControlFunctionFunctionalities` and has to support any object type, as any external consumer can use the same `register_pgn_request_callback`. I do agree it's not the greatest for readability and knowing whats happening, but it does provide a type agnostic way to get a pointer to an instance of the registrant, which is the main goal I was trying to achieve. Open to other ideas on this of course

Reply:
Ah I see, disregard my comment for now. We could maybe switch to the EventDispatcher system for the `register_pgn_request_callback` at some stage, but I'm not sure what side effects that might have or even performance loss.